provisioner/data.js

```diff
@@ -1,30 +1,704 @@
+     *   }
+     * } */
+    regions: base.Entity.types.JSON,
+  },
+  context: ['ec2', 'keyPrefix', 'pubKey', 'influx'],
 });
 
-// Export WorkerType
-exports.WorkerType = WorkerType;
\ No newline at end of file
+/**
+ * Create a workerType in the table.  The properties
+ * should not have a workerType key since that will be
+ * specified in the workerType argument
+ */
+WorkerType.create = function(workerType, properties) {
+  assert(workerType);
+  assert(properties);
+  assert(!properties.workerType);
+  properties.workerType = workerType;
+  return base.Entity.create.call(this, properties);
+};
+
+
+/**
+ * Return a list of all known workerTypes
+ */
+WorkerType.loadAll = function() {
+  var workers = [];
+
+  var p = base.Entity.scan.call(this, {}, {
+    handler: function (item) {
+      workers.push(item);
+    }
+  });
+
+  p = p.then(function() {
+    return workers;
+  });
+
+  return p;
+};
+
+/**
+ * Load the names of all known workerTypes
+ */
+WorkerType.listWorkerTypes = function() {
+  var names = [];
+
+  var p = base.Entity.scan.call(this, {}, {
+    handler: function (item) {
+      names.push(item.workerType);
+    }
+  });
+
+  p = p.then(function() {
+    return names;
+  });
+
+  return p;
+};
+
+
+/**
+ * Turn off every single EC2 instance and cancel all spot
+ * requests which were created by this Provisioner
+ */
+WorkerType.killEverything = function (debug) {
+  assert(debug);
+
+  var p = WorkerType.loadAll.call(this);
+
+  p = p.then(function (workerTypes) {
+    return Promise.all(workerTypes.map(function(workerType) {
+      return workerType.killAll(debug);
+    }));
+  });
+
+  return p;
+};
+
+
+/**
+ * Load a single workerType by name
+ */
+WorkerType.load = function(workerType) {
+  assert(workerType);
+  return base.Entity.load.call(this, {
+    workerType: workerType
+  });
+};
+
+
+/**
+ * Return an Object for JSON encoding which represents
+ * the data associated with this WorkerType.  This is a
+ * method intended for use in displaying the data associated
+ * with a given workerType
+ */
+WorkerType.prototype.json = function() {
+  return lodash.clone(this.__properties);
+};
+
+
+/**
+ * Return the list of regions which can be operated in.
+ * This is a subset of the regions which a given node
+ * is allowed to run in and the list of regions that the
+ * API is configred to run in.
+ */
+WorkerType.prototype.listRegions = function() {
+  var that = this;
+  return Object.keys(this.regions).filter(function(region) {
+    return that.ec2.regions.indexOf(region) !== -1;
+  }); 
+};
+
+
+/**
+ * We use KeyPair names to determine ownership and workerType
+ * in the EC2 world because we can't tag SpotRequests until they've
+ * mutated into Instances.  This sucks and all, but hey, what else
+ * can we do?  This method checks which regions have the required
+ * KeyPair already and creates the KeyPair in regions which do not
+ * already have it
+ */
+WorkerType.prototype.createKeyPair = function() {
```

Comment:
I would be tempted to move key management into `AwsState´ or do it in separate object. Mostly because it seems like it makes sense to refresh it every 20 min (perhaps) and just track state in-memory.

And key creation/deletion logic seems out of place here..
That say it might make sense to a add a utility function like `WorkerType.prototype.region()` that that returns a list of regions the workerType exists in... Abstracting away the `Object.keys()` and potential filtering.